### PR TITLE
Add VarInt Serializers [ECR-2380]

### PR DESCRIPTION
**Pull request by bullet-tooth**
## Overview
Implemented:
- Uint32
- Sint32
- Uint64
- Sint64

---
See: https://jira.bf.local/browse/ECR-2380

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [x] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [x] Public API has Javadoc
- [x] Method preconditions are checked and documented in the Javadoc of the method
- [x] Changelog is updated if needed (in case of notable or breaking changes)
- [x] The continuous integration build passes

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,7 +25,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   allowing to pass Protocol Buffer messages directly instead of using
   `StandardSerializers#protobuf`. (#505)
 - `StandardSerializers` now supports `bool`, `fixed32`, `fixed64`, `float` and `double` 
-  primitive types, `PrivateKey`, `PublicKey` and `bytes` serialization (#514)  
+  primitive types, `PrivateKey`, `PublicKey` and `bytes` serialization (#514)
+- `StandardSerializers` now supports `varInt32' and `varInt64` serialization (#523)
```
Comment (dmitry-timofeev) [inline CHANGELOG.md]: missing backtick

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java [hunk 2]
```diff
@@ -48,13 +48,27 @@
     return Fixed32Serializer.INSTANCE;
   }
 
+  /**
+   * Returns a serializer of integers using variable length encoding in little-endian byte order.
+   */
+  public static Serializer<Integer> varInt32() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java]: I’d use the naming from protobuf unless we support the only type — is it sint32/int32/uint32?

File: exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/VarInt32SerializerTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class VarInt32SerializerTest {
+
+  private Serializer<Integer> serializer = VarInt32Serializer.INSTANCE;
+
+  @ParameterizedTest
+  @ValueSource(ints = {Integer.MIN_VALUE, -1, 0, 1, Integer.MAX_VALUE})
+  void roundTrip(Integer value) {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/VarInt32SerializerTest.java]: I don't think this test is adequate — a more thorough input selection is needed (i.e., given how the implementation encodes various ranges of integers, pick some boundary values±1).

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java (hunk 2 above)
Comment (bullet-tooth) [reply exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java]: protobuf has 3 var-ints:
- `int32` -> from the doc: _Inefficient for encoding negative numbers – if your field is likely to have negative values, use sint32 instead._
- `uint32` -> stored as signed int in Java
- `sint32` -> uses uint32 encoding with `ZigZag32` encoding.

So I believe we need serializer for `sint32` only. That's because I gave `varInt32` name. And the same story for 64-bit ints.
@dmitry-timofeev @DENMROOT @MakarovS WDYT?

File: exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/VarInt32SerializerTest.java (hunk 3 above)
Comment (bullet-tooth) [reply exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/VarInt32SerializerTest.java]: changed source values

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java (hunk 2 above)
Comment (dmitry-timofeev) [reply exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java]: `uint32` is _represented_ as a signed int in Java, but I believe is encoded more efficiently than `sint32` because the latter performs zig-zag encoding (i.e., 127 (2^7 - 1) as `uint32` takes 1 byte; whilst the same as `sint32` takes 2 bytes): https://developers.google.com/protocol-buffers/docs/encoding#varints

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java [hunk 4]
```diff
@@ -63,10 +74,21 @@
...
 
   /**
-   * Returns a serializer of longs using variable length encoding in little-endian byte order.
+   * Returns a serializer of unsigned longs using variable length encoding.
+   * These more efficiently encodes values for the range {@code [0; 2^49-1]}
+   * than {@link StandardSerializers#fixed64()}.
    */
-  public static Serializer<Long> varInt64() {
-    return VarInt64Serializer.INSTANCE;
+  public static Serializer<Long> uint64() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/StandardSerializers.java]: Would you please use this in our QA service (for the `com.exonum.binding.qaservice.QaSchema#counters`)?

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java [hunk 5]
```diff
@@ -18,16 +18,17 @@
...
 import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Arrays.copyOf;
 
 import java.nio.BufferOverflowException;
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
 import java.util.Arrays;
 
-enum VarInt32Serializer implements Serializer<Integer> {
+enum UInt32Serializer implements Serializer<Integer> {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java]: It probably has to be `Uint32` because of the naming conventions.

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java [hunk 6]
```diff
@@ -62,7 +63,7 @@ public Integer fromBytes(byte[] serializedValue) {
       int x;
       if ((x = serializedValue[pos++]) >= 0) {
         return x;
-      } else if (serializedValue.length - pos < 9) {
+      } else if (serializedValue.length - pos < serializedValue.length - 1) {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java]: I am not sure what this condition means. It used to be `9`, but turned into (**a reminder to everyone: the type of the serializer _changed_ from `sint32` to `uint32`**)
```
serializedValue.length - pos < serializedValue.length - 1
  Strip duplicates:
- pos < - 1
  Keep going
1 < pos
  Swap
pos > 1
```

Is `pos > 1` the intended condition (please check my math)?

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt64Serializer.java [hunk 7]
```diff
@@ -57,27 +58,27 @@ public Long fromBytes(byte[] serializedValue) {
...
-      int tempPos = 0;
+      int pos = 0;
 
       long x;
       int y;
-      if ((y = serializedValue[tempPos++]) >= 0) {
+      if ((y = serializedValue[pos++]) >= 0) {
         return (long) y;
-      } else if (serializedValue.length - tempPos < 9) {
+      } else if (serializedValue.length - pos < serializedValue.length - 1) {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt64Serializer.java]: Same as in UInt32.

File: exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/SInt32SerializerTest.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+  private static List<byte[]> invalidVarInts() {
+    return ImmutableList.of(
+        Bytes.bytes(),
+        Bytes.bytes(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11)
+    );
+  }
+
+  @Test
+  void name() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/SInt32SerializerTest.java]: This one probably must be removed

File: CHANGELOG.md [hunk 9]
```diff
@@ -25,8 +25,8 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
   allowing to pass Protocol Buffer messages directly instead of using
   `StandardSerializers#protobuf`. (#505)
 - `StandardSerializers` now supports `bool`, `fixed32`, `fixed64`, `float` and `double` 
-  primitive types, `PrivateKey`, `PublicKey` and `bytes` serialization (#514)
-- `StandardSerializers` now supports `varInt32' and `varInt64` serialization (#523)    
+  primitive types, `PrivateKey`, `PublicKey` and `bytes` serialization. (#514)
+- `StandardSerializers` now supports `uint32`, `sint32`, `uint64` and `sint64` serialization. (#523)
```
Comment (dmitry-timofeev) [inline CHANGELOG.md]: I'd consider joining the lines.

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java (hunk 6 above)
Comment (bullet-tooth) [reply exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java]: `9` came from the original code and it means `MAX_BYTES -1`.  Such as we need to remove unused space tail in the buffer we can't use `MAX_BYTES` here.
Yes, I'll simplify this statement.

Comment (dmitry-timofeev) [reply exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java]: > Such as we need to remove unused space tail in the buffer we can't use MAX_BYTES here.

Not sure I understand — do we remove anything? When this condition can possibly be true (we only `break` out of the block, never `continue`)? What the original code of uint32 serializer uses?

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+      } else if ((x ^= (serializedValue[pos++] << 14)) >= 0) {
+        x ^= (~0 << 7) ^ (~0 << 14);
+      } else if ((x ^= (serializedValue[pos++] << 21)) < 0) {
+        x ^= (~0 << 7) ^ (~0 << 14) ^ (~0 << 21);
+      } else {
+        int y = serializedValue[pos++];
+        x ^= y << 28;
+        x ^= (~0 << 7) ^ (~0 << 14) ^ (~0 << 21) ^ (~0 << 28);
+        if (y < 0
+            && serializedValue[pos++] < 0
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/UInt32Serializer.java]: At this moment `pos == 5` (i.e., it attempts to access the array out of its bounds). This code does not seem correct for **u**int32.

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/SerializationUtils.java [hunk 11]
```diff
@@ -33,6 +36,16 @@ static void checkLength(byte[] array, int length) {
         "Expected an array of size %s, but was %s", length, array.length);
   }
 
+  /**
+   * Performs check that the array has no bytes after the position.
+   *
+   * @throws IllegalArgumentException thrown if any bytes in the tail exist
+   */
+  static void checkNoTailLeft(byte[] array, int pos) {
+    checkArgument(pos == array.length, "Expected no tail left in the array, but was %s",
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/SerializationUtils.java]: Probably, checkArrayLength would have been enough, but us we provide extra
information, the expected size (which `pos` actually is) must be included in
the error message.

File: exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java [hunk 12]
```diff
@@ -24,4 +30,55 @@ static void invalidBytesValueTest(byte[] invalidValue, Serializer serializer) {
     assertThrows(IllegalArgumentException.class, () -> serializer.fromBytes(invalidValue));
   }
 
+  static IntStream intValues() {
+    return IntStream.range(0, 32)
+        .map(value -> 1 << value)
+        .flatMap(value -> IntStream.of(-value, value - 1, value, value + 1));
+  }
+
+  private static LongStream longValues() {
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java]: The methods used as sources _in other classes_ must not be private — that breaks IDEA (they are shown as unused).

File: exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java [hunk 13]
```diff
@@ -24,4 +30,55 @@ static void invalidBytesValueTest(byte[] invalidValue, Serializer serializer) {
...
+        Bytes.bytes(),
+        // # MSB is set in the last byte, but there is the end of serialized value:
+        Bytes.bytes(0x80),
+        Bytes.bytes(0x8F),
+        Bytes.bytes(0xFF),
+        Bytes.bytes(0x80, 0x81),
+        Bytes.bytes(0x80, 0x81, 0x82),
+        Bytes.bytes(0x80, 0x81, 0x82, 0x83),
+        Bytes.bytes(0x80, 0x81, 0x82, 0x83, 0x84),
+        Bytes.bytes(0x80, 0x81, 0x82, 0x83, 0x84, 0x85),
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java]: Values for longs (probably doesn't make sense to add them to 32-bit)?

File: exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java [hunk 14]
```diff
@@ -24,4 +30,55 @@ static void invalidBytesValueTest(byte[] invalidValue, Serializer serializer) {
...
+
+        // # Correct first bytes, but unexpected "tail":
+        Bytes.bytes(0x01, 0x02), // A single byte varint + tail.
+        Bytes.bytes(0x01, 0x82), // A single byte varint + tail.
+        Bytes.bytes(0x01, 0x02, 0x03, 0x04, 0x85), // A single byte varint + 4-byte tail.
+        Bytes.bytes(0x80, 0x81, 0x02, 0x83, 0x84), // A valid 3 byte varint + 2-byte tail
+        Bytes.bytes(0x80, 0x81, 0x82, 0x03, 0x84), // A valid 4 byte varint, 5th byte is invalid
+        Bytes.bytes(0x80, 0x81, 0x82, 0x83, 0x04, 0x01) // Valid 5 byte varint, 6th byte is invalid
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java]: long-tails? :cocktail:

File: exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/Uint32Serializer.java [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+        x ^= (~0 << 7) ^ (~0 << 14);
+      } else if ((x ^= (serializedValue[pos++] << 21)) < 0) {
+        x ^= (~0 << 7) ^ (~0 << 14) ^ (~0 << 21);
+      } else {
+        int y = serializedValue[pos++];
+        x ^= y << 28;
+        x ^= (~0 << 7) ^ (~0 << 14) ^ (~0 << 21) ^ (~0 << 28);
+        if (y < 0) {
+          throw new IllegalArgumentException(
+              "Serialized value has wrong format " + Arrays.toString(serializedValue));
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/main/java/com/exonum/binding/common/serialization/Uint32Serializer.java]: I'd clarify a bit the error message:
"Serialized value has wrong format, the last (5th) byte has MSB set."

File: exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java [hunk 16]
```diff
@@ -53,8 +53,11 @@ static LongStream longValues() {
 
   static Stream<byte[]> invalidVarints64() {
     return concat(malformedVarints(), Stream.of(
+        // # MSB is set in the last byte, but there is the end of serialized value:
+        Bytes.bytes(0x80, 0x81, 0x82, 0x83, 0x84, 0x85),
```
Comment (dmitry-timofeev) [inline exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java]: Up to 10 bytes?

Comment (bullet-tooth) [reply exonum-java-binding-common/src/test/java/com/exonum/binding/common/serialization/StandardSerializersTest.java]: done
